{{cookiecutter.project_name}}/.azure/devops-pipelines/tests-ci.yml

```diff
@@ -0,0 +1,130 @@
+
+    steps:
+    - checkout: self
+      persistCredentials: true
+      clean: true
+      displayName: 'Checkout & Build.Reason: $(Build.Reason) & Build.SourceBranchName: $(Build.SourceBranchName)'
+
+    - task: UsePythonVersion@0
+      displayName: 'Use Python 3.8'
+      inputs:
+        versionSpec: 3.8
+
+    - script: |
+        python -m pip install --upgrade pip
+        pip install databricks-cli
+      displayName: 'Install dependencies'
+
+    - script: |
+        set -e
+        STAGING_AZURE_SP_TENANT_ID=$(STAGING_AZURE_SP_TENANT_ID)
+        STAGING_AZURE_SP_APPLICATION_ID=$(STAGING_AZURE_SP_APPLICATION_ID)
+        STAGING_AZURE_SP_CLIENT_SECRET=$(STAGING_AZURE_SP_CLIENT_SECRET)
+        DATABRICKS_TOKEN=$(.azure/devops-pipelines/scripts/generate-aad-token.sh "$STAGING_AZURE_SP_TENANT_ID" "$STAGING_AZURE_SP_APPLICATION_ID" "$STAGING_AZURE_SP_CLIENT_SECRET")
+        echo "##vso[task.setvariable variable=DATABRICKS_TOKEN;issecret=true]${DATABRICKS_TOKEN}"
+      displayName: 'Configure AAD auth'
+
+    - script: |
+        if [[ "$(Build.Reason)" == "PullRequest" ]]
+        then
+          pr_source_branch=$(System.PullRequest.SourceBranch)
+          GIT_BRANCH=${pr_source_branch#refs/heads/}
+        else
+          ORIGINAL_BRANCH='$(Build.SourceBranch)'
+          ROOT=refs/heads/
+          GIT_BRANCH="${ORIGINAL_BRANCH/$ROOT/''}"
+        fi
+        databricks jobs configure --version=2.1
+        databricks runs submit --wait --json \
+          '{
+            "run_name": "integration-test",
+            "tasks": [
+              {
+                "task_key": "training",
+                "notebook_task": {
+                  "notebook_path": "notebooks/Train",
+                  "base_parameters": {
+                    "env": "staging",
+                    "test_mode": "True"
+                    }
+                  },
+                "new_cluster": {
+                  "node_type_id": "Standard_D3_v2",
+                  "spark_version": "11.2.x-cpu-ml-scala2.12",
+                  "num_workers": 0,
+                  "spark_conf": {
+                    "spark.master": "local[*,4]",
+                    "spark.databricks.cluster.profile": "singleNode"
+                    },
+                  "custom_tags": {
+                    "ResourceClass": "SingleNode"
+                    }
+                  }
+                }
+              ],
+            "git_source": {
+              "git_url": "$(Build.Repository.Uri)",
+              "git_branch": "'"$GIT_BRANCH"'",
+              "git_provider": "azureDevOpsServices"
+              }
+          }'
+      env:
+        DATABRICKS_HOST: $(STAGING_DATABRICKS_HOST)
+        DATABRICKS_TOKEN: $(DATABRICKS_TOKEN)
+      displayName: 'Trigger integration test'
+
+    - task: PublishTestResults@2
+      condition: succeededOrFailed()
+      inputs:
+        testResultsFormat: 'JUnit'
+        testResultsFiles: '**/test-*.xml'
+        failTaskOnFailedTests: true
```

Comment:
I wonder if there's any easy way to dedupe this job config? I think it's the same as the one we use for the github actions integration test. @smurching

Reply:
Indeed it is. We could pull it out into a common shell script and run it in GitHub Actions too. Longer term, our Jobs team will improve the story here for testing and deploying jobs (note that even if we pull this out into a shell script shared with GitHub Actions, the cluster spec used for the job is currently duplicated with what's in Terraform)

Reply:
I don't think it's urgent to update our GitHub Actions YAMLs to use the Databricks CLI (we'd lose some of the readability benefits of using the native https://github.com/databricks/run-notebook Action to run integration tests), but it would be nice to pull the `databricks runs submit` invocation here into a shell script so we can reuse it across other CI/CD providers for which we don't have an analog for https://github.com/databricks/run-notebook

Reply:
I had actually started to try to do something similar to the above suggestion. I looked at pulling the common cluster config out, and have a `payload_cluster_config.json` with the configs for a new cluster, but it started to look a bit overly complex. Specifically for passing that json into the task in the Azure testing-ci.yml script

Reply:
I see - agreed, I think this would be best solved w/ "workflows as code". We can reference the (CI/CD provider-agnostic) workflow configs from GH Actions, ADO, Terraform and other places.

Reply:
Agreed with Sid for supporting other git providers it would then make sense to pull out into a shell script. Until we have introduced those other providers I would vote to keep this as is for now